src/math/exp_fcn.c

```diff
@@ -0,0 +1,164 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Shriram Shastry <malladi.sastry@linux.intel.com>
+//
+//
+
+/* Include Files */
+#include <sof/math/exp_fcn.h>
+#include <stdint.h>
+#include <stdbool.h>
+
+/* Function Declarations */
+static inline int64_t lomul_s64_SR_sat_near(int64_t a, int64_t b);
+
+static inline void wide_mul_s64(int64_t in_0, int64_t in_1, uint64_t *bitshiptrout,
+				uint64_t *bitsloptrout);
+/*
+ * Arguments	: int64_t a
+ *		  int64_t b
+ * Return Type	: int64_t
+ */
+static inline int64_t lomul_s64_SR_sat_near(int64_t a, int64_t b)
+{
+	uint64_t u64_rhi;
+	uint64_t u64_rlo;
+	bool roundup;
+
+	wide_mul_s64(a, b, &u64_rhi, &u64_rlo);
+	roundup = ((u64_rlo & BIT_MASK_LOW_Q27P5) != 0ULL);
+	u64_rlo =
+	    (u64_rhi << 36 | u64_rlo >> 28) + (roundup ? (uint64_t)1 : (uint64_t)0);
```

Comment:
What is the issue with the code having a ternary that you see?
If you wanted to propose a cycle-based recommendation, I don't see any noticeable improvement there. 

```
		|cycMax	|cycMin|
----------------+-------|------+
suggestion - 1: |1348   |   191|
suggestion - 2:	|1388   |   197|
suggestion - 3:	|1388   |   197|
Nativecode - 1:	|1348   |   191|


```

Reply:
how do you count those cycles? You have that code run with different input data? How come the range is so huge for such a simple and deterministic code? Was that with optimisation?
The issue is that the ternary operation is a hidden `if` and an `if` (unless optimised out by the compiler) generates a jump, and jumps can be bad for performance

Reply:
_how do you count those cycles?_
I count cycles by reading cycles from a cycle register. 

_You have that code run with different input data ?_
There are no brand-newer facts introduced in addition.
The results  are from 128 samples of data in the -5 to + 5 range, so I've included the minimum and maximum cycle counts. 

_How come the range is so huge for such a simple and deterministic code?_ 
If you'd like me to share the address and instructions for looking within the profile disassembly code for the aforementioned three proposals, please let me know.  

_Was that with optimisation?_
Results come from the same configuration; optimization settings haven't changed.

_The issue is that the ternary operation is a hidden if and an if (unless optimised out by the compiler) generates a jump, and jumps can be bad for performance_

I'm unsure and don't understand what you're saying. Although the compiler hasn't changed for the aforementioned measurement, you can observe that the cycle counts for suggestions 1 and 2 are different.

Reply:
instead of provide min and max make people confusing, better to provide cycles for e^0.12345, e^(-0.12345) e^(1.12345) e^(-1.12345) such data, developer may have a rough picture for this new exp function.

This function mainly extend the range for exp compared with previous small one.

Reply:
Thank you - Good idea


If you look at the `exp int32()`, plot https://github.com/thesofproject/sof/pull/6655#issuecomment-1325867167 you'll notice that it has a V shape, meaning that resources and cycles are less while the numeric input is tiny and steadily increase as the numeric value increases.


If the cycle count is inconsistent, the implementation of the proposed `exp_int32()` algorithm is flawed. 

_This function mainly extend the range for exp compared with previous small one._
Yes,  the `exp_int32()` function cycles are also smaller 

You can get a good notion of `exp int32()` works by comparing it to comparisons between Hifi3, Hifi4, and Hifi5 Generic compilers using the -1.9/+1.9 fixed numeric

```
		                HiFi3
		     Input	Output	  cycle 	
		------------+------------+------+
Builtin exp()	 -1.92913386|  0.14527397| 14689|
		 1.92913386 |  6.88354553| 14671| 	
		
				HiFi3
		    Input	Output	  cycle 	
		-----------+-------------+-----+							
exp_int32()     -1.92913386|   0.14527397| 1455|				
		1.92913386 |   6.88354553| 1381|
		
				HiFi3
		   Input	Output	     cycle 	
		-------------+-------------+------+				
exp_small_fixed	-1.9928595275| 0.1362981796|  9406|	
                 1.9928595275| 7.3364725113|  9310|
		 
			         HiFi4						 
		   Input	Output	      cycle 	
		--------------+-------------+------+	
exp_small_fixed	 -1.9928595275| 0.1362981796|  958|  
		  1.9928595275| 7.3364725113|  958|  
		  
				HiFi4						 
		   Input	Output	   cycle 			 
		-----------+-- ----------+------+		 
exp_int32()	-1.92913386|   0.14526463|  867|	
		 1.92913386|   6.88354230|  858|	

				HiFi4				 
		      Input	Output	  cycle 	
		-----------+-------------+-----+
Builtin exp()   -1.92913386|   0.14527397| 1859|	
		1.92913386 |   6.88354553| 1884|	
		
		
				HiFi5		
		         Input	 Output	        cycle
		--------------+---------------+-------+
exp_small_fixed  -1.9928595275|   0.1362981796|    878|
		  1.9928595275|   7.3364725113|    878|
			
				HiFi5		
		    Input 	Output		cycle
	       ------------+---------------+----------+
exp_int32()	-1.92913386|     0.14526463|       737|
		 1.92913386|	 6.8835423 |       727|

				HiFi5		
		    Input    	Output    	 cycle
	       ------------+---------------+---------+
Builtin exp()	-1.92913386|     0.14527397|     1847|
		 1.92913386|     6.88354553|     1871| 	


```

Reply:
> instead of provide min and max make people confusing, better to provide cycles for e^0.12345, e^(-0.12345) e^(1.12345) e^(-1.12345) such data, developer may have a rough picture for this new exp function.
> 
> This function mainly extend the range for exp compared with previous small one.

The existing exp function is two parts, the static exp_small_fixed() that is Taylor series with 11 first terms.  It handles range -2 to 2. The second function that is for client apps is exp_fixed() with input range -11.5 to 7.6. @ShriramShastry please compare your work to exp_fixed and not the small values function. It's static, so it can't be used by other code.

Can you please share in a clear understandable way the MCPS and accuracy of this work vs. exp_fixed() to help decide if we want to keep both and per case select which exponent to use or replace old version with this.